### ISSUE TITLE
Cannot pass value for a config key from commands

### ISSUE THREAD

**Issue by xuanswe** [pulumi/pulumi#3373]
Terraform support input variables with interactive [ui input](https://learn.hashicorp.com/terraform/getting-started/variables.html#ui-input) when the variable has no value.

I know that Pulumi support [config](https://www.pulumi.com/docs/intro/concepts/config/). But there are 2 missing features when using pulumi config when compare to terraform input variables.
1. Cannot pass a value from commands
Terraform: `terraform apply -var="key1=value1" -var="key2=value2"`
Expected pulumi has smt like this: `pulumi up --config-key1=value1 --config-key2=value2`
2. Do not ask user to input requred value interactively
Terraform:
```
$ terraform apply
var.server_port
  The port the server will use for HTTP requests
  Enter a value:
```
Expected pulumi has smt like this:
```
$ pulumi up
config.serverPort
  The port the server will use for HTTP requests
  Enter a value:
```

**Comment by mikhailshilkov** [comment]
You can pass config values with the `--config stringArray` parameter, see https://www.pulumi.com/docs/reference/cli/pulumi_up/#options

**Comment by nguyenxndaidev** [comment]
@mikhailshilkov the behavior of `--config stringArray` is to save value to `Pulumi.<stack>.yaml` file. That is not so useful in case you want ad hoc value.
The value in `Pulumi.<stack>.yaml` should be considered as default value. The value provided in command line should be used and shouldn't save to the yaml file.

**Comment by leezen** [comment]
Closing out old issues. I would say this is "by design" as we want to persist this configuration value in the state.

**Comment by cloudspeak** [comment]
I would also like to have this feature so wanted to share my use case in the hope it helps.  Our build pipeline allows Pulumi to work against local mocks of certain AWS services, so we launch these servers and tell Pulumi which URL and port to use.  But these values are transient, because they are only valid as long as the servers are running.  The next time they run they could be on different hosts or ports.  Saving this type of config to a file clutters the codebase with outdated config and makes source control messy; and we don't want to .gitignore the file because it contains other config which is important.  I can imagine there are other such situations where state is only temporarily valid.

**Comment by leezen** [comment]
@cloudspeak In the case you described, those values don't sound like "stack config" to me? In other words, you don't need to rely on the stack config mechanism to pass those values in. Instead, couldn't you use an environment variable?

**Comment by cloudspeak** [comment]
@Leezen True, I wouldn't really configure it "stack config" as such, more "parameters".  I think both @nguyenxndaidev and I are looking for something more elegant than setting lots of global environment variables to pass values in.  Not essential, but a nice to have.

**Comment by robertwe** [comment]
Very similar solution is possible in Terraform and I must admit it's useful in some cases especially when u want to disable/enable specific block. Such feature is also give us some space to maneuver on the CI/CD level.

**Comment by nguyenxndaidev** [comment]
> Closing out old issues. I would say this is "by design" as we want to persist this configuration value in the state.

If the design doesn't reflex the real-life requirements, it should be redesign. IMO, this ticket should be reopen.
Honestly, I don't know what's wrong if this feature is supported, especially, it's working very well with Terraform.

**Comment by ghostsquad** [comment]
@leezen can you reopen this please?

**Comment by leezen** [comment]
I like the suggestion to allow passing additional args (vs. config) and letting the program decide what to do with them.

**Comment by treykasada** [comment]
A potential approach for this might be environment variable templating in the stack configuration:
```yaml
config:
  aws:region: '{{.AWS_REGION | default "ap-southeast-2"}}'
```
(the above example assumes Go template syntax, but other syntaxes like Bash could also work)
